xde/src/xde.rs

```diff
@@ -908,47 +952,67 @@ fn delete_xde(req: &DeleteXdeReq) -> Result<NoResp, OpteError> {
         }
     };
 
-    // Remove the xde device entry.
-    _ = devs.remove(&req.xde_devname);
     Ok(NoResp::default())
 }
 
+struct ResolvedLink<'a>(&'a str, LinkId);
+impl<'a> ResolvedLink<'a> {
+    fn new(name: &'a str) -> Result<Self, OpteError> {
+        let link_cstr = CString::new(name).unwrap();
+
+        let link_id =
+            LinkId::from_name(link_cstr).map_err(|err| OpteError::System {
+                errno: EFAULT,
+                msg: format!("failed to get linkid for {name}: {err}"),
+            })?;
+
+        Ok(Self(name, link_id))
+    }
+}
+
 #[unsafe(no_mangle)]
 fn set_xde_underlay(req: &SetXdeUnderlayReq) -> Result<NoResp, OpteError> {
     let state = get_xde_state();
 
-    let mut underlay = state.underlay.lock();
-    if underlay.is_some() {
+    let mut token = state.management_lock.lock();
+
+    if token.underlay.is_some() {
         return Err(OpteError::System {
             errno: EEXIST,
             msg: "underlay already initialized".into(),
         });
     }
-    *underlay = Some(unsafe {
-        init_underlay_ingress_handlers(req.u1.clone(), req.u2.clone())?
-    });
+
+    // Resolve `LinkId`s outside of any locks -- these require upcalls.
+    let link1 = ResolvedLink::new(req.u1.as_str())?;
+    let link2 = ResolvedLink::new(req.u2.as_str())?;
```

Comment:
Does LinkId resolution need to happen under the token lock, or could it be done before?